drivers/greth.c

```diff
@@ -0,0 +1,564 @@
+	size_t sz;
+	if ((desc->flags & GRETH_DESC_EN) != 0) {
+		return 0;
+	}
+	sz = (desc->flags >> GRETH_DESC_LEN_SHIFT) & GRETH_DESC_LEN_MASK;
+	if (sz == 0) {
+		sz = 1;
+		printf("lwip: greth: WARNING: This message indicates a potential HW bug:\n");
+		printf("lwip: greth: HW provided invalid size: %zu. Setting size to 1\n", sz);
+	}
+	return sz;
+}
+
+
+static int greth_pktRxFinished(const net_bufdesc_ring_t *ring, size_t i)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+
+	return !(desc->flags & GRETH_DESC_EN);
+}
+
+
+static void greth_fillRxDesc(const net_bufdesc_ring_t *ring, size_t i, addr_t pa, size_t sz, unsigned seg)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	unsigned wrap = desc == (volatile greth_buf_desc_t *)ring->ring + ring->last ? GRETH_DESC_WR : 0;
+
+	desc->addr = pa;
+
+	atomic_thread_fence(memory_order_seq_cst);
+	desc->flags = GRETH_DESC_EN | GRETH_DESC_IE | wrap;
+	atomic_thread_fence(memory_order_seq_cst);
+}
+
+
+static int greth_nextTxDone(const net_bufdesc_ring_t *ring, size_t i)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	return !(desc->flags & GRETH_DESC_EN);
+}
+
+
+static void greth_fillTxDesc(const net_bufdesc_ring_t *ring, size_t i, addr_t pa, size_t sz, unsigned seg)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	uint32_t flags;
+
+	flags = GRETH_DESC_EN;
+	flags |= (i == ring->last) ? GRETH_DESC_WR : 0;
+	flags |= (sz << GRETH_DESC_LEN_SHIFT) & GRETH_DESC_LEN_MASK; /* shift (0) just for consistency  */
+	flags |= ((seg & BDRING_SEG_LAST) != 0) ? 0 : GRETH_DESC_TX_MO;
+	flags |= GRETH_DESC_IE;
+
+#if GRETH_CHECKSUM_OFFLOAD
+	flags |= GRETH_DESC_TX_UC | GRETH_DESC_TX_TC | GRETH_DESC_TX_IC;
+#endif
+
+	desc->addr = pa;
+	atomic_thread_fence(memory_order_seq_cst);
+	desc->flags = flags;
+	atomic_thread_fence(memory_order_seq_cst);
+}
+
+
+static const net_bufdesc_ops_t greth_ring_ops = {
+	.nextRxBufferSize = greth_nextRxBufferSize,
+	.pktRxFinished = greth_pktRxFinished,
+	.fillRxDesc = greth_fillRxDesc,
+	.nextTxDone = greth_nextTxDone,
+	.fillTxDesc = greth_fillTxDesc,
+	.desc_size = sizeof(greth_desc_t),
+	.pkt_buf_sz = GRETH_MAX_PKT_SZ,
+	.ring_alignment = 1024,
+	.max_tx_frag = 0xFFFF,
+};
+
+static const size_t greth_ring_sizes[] = { GRETH_RX_RING_SIZE, GRETH_TX_RING_SIZE };
+
+
+static int greth_initRings(greth_state_t *state)
+{
+	int err;
+
+	err = net_initRings(state->rings, greth_ring_sizes, sizeof(state->rings) / sizeof(*state->rings), &greth_ring_ops);
+	if (err != 0) {
+		return err;
+	}
+
+	state->mmio->TX_DESC_PTR = state->tx.phys;
+	state->mmio->RX_DESC_PTR = state->rx.phys;
+
+	return 0;
+}
+
+
+static err_t greth_netifOutput(struct netif *netif, struct pbuf *p)
+{
+	greth_state_t *state = netif->state;
+	size_t nf;
+
+	if (ETH_PAD_SIZE != 0) {
+		(void)pbuf_header(p, -ETH_PAD_SIZE); /* drop the padding word */
```

Comment:
Maybe comment about why we're doing that
```suggestion
		(void)pbuf_header(p, -ETH_PAD_SIZE); /* drop the padding word (not supported by HW) */
```

Reply:
also: `ETH_PAD_SIZE` is known at compile time, so you should check it with preprocessor

if there's anything similar in other drivers, please fix it also

Reply:
done, added comment and changed it to compile-time check